diff --git a/CMakeLists.txt b/CMakeLists.txt
index e04cefd..e26bd32 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -102,6 +102,9 @@ if(WIN32)
         target_sources(SDL2_net PRIVATE version.rc)
     endif()
 endif()
+if(UNIX AND NOT ANDROID)
+    target_link_libraries(SDL2_net PRIVATE socket)
+endif()
 set_target_properties(SDL2_net PROPERTIES
     DEFINE_SYMBOL DLL_EXPORT
     EXPORT_NAME ${sdl2_net_export_name}
